### PR TITLE
Change to new C++ based addon interface style

### PR DESCRIPTION
**Pull request by Rechi**
This are changes related to Kodi request xbmc/xbmc#12262

@AlwinEsch please check if the new interface is used correct.

### CODE REVIEW

File: lib/vis_milkdrop/utility.cpp [hunk 1]
```diff
@@ -346,7 +345,7 @@ void GetFromCache(char *cache, char *key, char *szDefault, char *buffer, int buf
 
 void InternalGetPrivateProfileSection(char *szSectionName, char *buffer, int size, char *szIniFile)
 {
-	void* handle;
+	kodi::vfs::CFile handle = kodi::vfs::CFile();
```
Comment (AlwinEsch) [inline lib/vis_milkdrop/utility.cpp]: A `kodi::vfs::CFile handle;` is enough

File: lib/vis_milkdrop/utility.cpp [hunk 2]
```diff
@@ -356,17 +355,15 @@ void InternalGetPrivateProfileSection(char *szSectionName, char *buffer, int siz
 
 	validIni = false;
 
-	handle = KODI->OpenFile(szIniFile, 0);
-	
-	if (handle == 0)
+	if (handle.OpenFile(szIniFile))
 	{
 		return;
 	}
```
Comment (AlwinEsch) [inline lib/vis_milkdrop/utility.cpp]: Here you have a mistake, in case of error returns OpenFile false

File: src/MilkdropXBMC.cpp [hunk 3]
```diff
@@ -118,59 +131,52 @@ void urlEscape(std::string& str)
...
+  if (!g_plugin)
+  {
+    g_plugin = new CPlugin;
+    g_plugin->PluginPreInitialize(0, 0);
+  }
+
+  if (!g_plugin || !g_plugin->PluginInitialize(static_cast<ID3D11DeviceContext*>(Device()), X(), Y(), Width(), Height(), PixelRatio()))
     return ADDON_STATUS_UNKNOWN;
 
   return ADDON_STATUS_NEED_SAVEDSETTINGS; // We need some settings to be saved later before we quit this plugin
```
Comment (AlwinEsch) [inline src/MilkdropXBMC.cpp]: This muste be changed to ADDON_STATUS_OK and related settings readed here.

File: src/MilkdropXBMC.cpp [hunk 4]
```diff
@@ -118,59 +131,52 @@ void urlEscape(std::string& str)
...
 }
 
-extern "C" void Start(int iChannels, int iSamplesPerSec, int iBitsPerSample, const char* szSongName)
-{}
+bool CVisualizationMilkdrop::Start(int iChannels, int iSamplesPerSec, int iBitsPerSample, std::string szSongName)
+{
+  return true;
+}
```
Comment (AlwinEsch) [inline src/MilkdropXBMC.cpp]: If you not need it can it completely stay away. But can also be leaved if needed in future for improvement.

**Comment by AlwinEsch** [comment]
Except the few things it looks good.

**Comment by Rechi** [comment]
@AlwinEsch updated, please have a look again

File: lib/vis_milkdrop/plugin.cpp [hunk 5]
```diff
@@ -4766,19 +4762,19 @@ void CPlugin::UpdatePresetList()
 		m_nPresets = 0;
 		m_nDirs    = 0;
 
-		if(numItems)
+		if(items.size())
```
Comment (AlwinEsch) [inline lib/vis_milkdrop/plugin.cpp]: You can also use a `if(!items.empty())` but can be also leaved like them.

**Comment by Rechi** [comment]
@AlwinEsch added now load settings during addon create

**Comment by afedchin** [comment]
thanks a lot

**Pull request merged**